ledger/voters.go

```diff
@@ -174,39 +200,66 @@ func (vt *votersTracker) newBlock(hdr bookkeeping.BlockHeader) {
 		return
 	}
 
-	vt.removeOldVoters(hdr)
-
 	// This might be a block where we snapshot the online participants,
 	// to eventually construct a vector commitment in a later
 	// block.
-	r := uint64(hdr.Round)
-	if (r+proto.StateProofVotersLookback)%proto.StateProofInterval != 0 {
+	r := hdr.Round
+	if (uint64(r)+proto.StateProofVotersLookback)%proto.StateProofInterval != 0 {
 		return
 	}
 
-	_, ok := vt.votersForRoundCache[basics.Round(r)]
-	if ok {
+	_, exists := vt.getVoters(r)
+	if exists {
 		vt.l.trackerLog().Errorf("votersTracker.newBlock: round %d already present", r)
 	} else {
 		vt.loadTree(hdr)
 	}
 
 }
 
+func (vt *votersTracker) prepareCommit(dcc *deferredCommitContext) error {
+	vt.commitListenerMu.RLock()
+	defer vt.commitListenerMu.RUnlock()
+
+	if vt.commitListener == nil {
+		return nil
+	}
+
+	commitListener := vt.commitListener
+	vf := votersFetcher{vt: vt}
+	// In case the listener's function fails, we do not want to break the commit process.
+	// To implement this hierarchy we've decided to not include a return value in OnPrepareVoterCommit function
+	commitListener.OnPrepareVoterCommit(dcc.oldBase, dcc.newBase(), &vf)
```

Comment:
this callback is still the most concerning part of the entire implementation.
just an idea... maybe make a subpackage `stateproof/data` where to define algorithm and data needed for a single builder like voters at round X, range of block headers. This function prepares all the data needed and the listener (stateproof builder) gets notified. Since it already have all data collected, it does not depends on ledger, and can create the builder asynchronously, so builder does not call `BlockHdr` and `VotersForStateProof`, nether blocks the goroutine.  Since it is non-blocking, this could be done with `OnNewBlock` notification instead of `OnPrepareVoterCommit`. Not sure if it good or bad proposal, so opinions? @algoidan @algonautshant

Reply:
I share @algorandskiy 's concerns. 
In addition to the overall structural complexity of this call, I also worry about the expensive operation made synchronously on `prepareCommit`, which may delay subsequent trackers' `prepareCommit` calls.

As I understand from the description of the proposal here, there will be a service continuously building stateproofs, and be independent of the ledger. The ledger will independently park the data necessary for the stateproof building, and notify the builder service about the progress in the ledger. 
I will add a context to that service as well, so we know when to start and stop (catchpoint catchup etc.)

Reply:
@algorandskiy, can you open a ticket please?